vernac/metasyntax.ml

```diff
@@ -477,6 +477,9 @@ let warn_format_break =
          (fun () ->
           strbrk "Discarding format implicitly indicated by multiple spaces in notation because an explicit format modifier is given.")
 
+let has_ldots l =
+  List.exists (function (_,UnpTerminal s) -> String.equal s (Id.to_string Notation_ops.ldots_var) | _ -> false) l
```

Comment:
This comment is not directly related to the PR, and I don't really understand this code, so sorry if this question is stupid, but: why on earth don't we use a richer ADT to represent reserved tokens for `UnpTerminal`? Here and in several other places in metasyntax, we hardwire the use of `..` when we could have used another constructor. Similar remark for other reserved clusters, like blank tokens.

Reply:
I agree with your comments on the principle but it is basically a compromise. The type defining `UnpTerminal` was introduced (I think) by Chet. Recursive notations were added in a second time. The type was concise and rather general and I added the interpretation of `..` as an extra parsing step on top of the original step. I could combine the two steps into one but this would make the parsing function globally more intricate. I could also use an intermediate data-structure, say:
```
type flat_unparsing 
  | FlatUnpMetaVar of int * parenRelation
  | FlatUnpBinderMetaVar of int * parenRelation  
  | FlatUnpTerminal of string
  | FlatUnpRecursiveKey                               
  | FlatUnpBox of ppbox * flat_unparsing Loc.located list
  | FlatUnpCut of ppcut
```
which would be used as an intermediate step in the process of interpreting a notation into the full structured unparsing type:
```
type unparsing =   
  | UnpMetaVar of int * parenRelation    
  | UnpBinderMetaVar of int * parenRelation      
  | UnpListMetaVar of int * parenRelation * unparsing list  
  | UnpBinderListMetaVar of int * bool * unparsing list 
  | UnpTerminal of string      
  | UnpBox of ppbox * unparsing Loc.located list   
  | UnpCut of ppcut  
```

I made https://github.com/herbelin/github-coq/pull/new/master+isolate-ldots_var-treatment to isolate the part of the code which uses `..` in `symbol` (i.e. the `Terminal` parsing part, not the `UnpTerminal` printing part), hoping that it becomes clearer that this is only a local hack. Tell me if you think it is worth to be continued, e.g. by merging the `interp_list_parser` with `decompose_raw_notation` so as to avoid the intermediate step temporarily mentioning `Terminal ".."`.

Reply:
I suspect that a lot of what can be found in Metasyntax and related is due to accidental complexity rather than actual tricky behaviours. Now that we have control over gramlib, it might be worthwhile to think about a thorough rewriting of that part of the code.

Reply:
As a general rule, I also believe that the complexity of `Metasyntax` could be reduced. And as a general rule, I also believe that the control over gramlib will help simplifying tricks only used to satisfy the communication with camlp5. Feel free to rewrite/rethink the way the code is done.